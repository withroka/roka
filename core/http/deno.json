{
  "name": "@roka/http",
  "version": "0.2.2",
  "exports": {
    ".": "./http.ts",
    "./json/client": "./json/client.ts",
    "./graphql/client": "./graphql/client.ts",
    "./request": "./request.ts",
    "./testing": "./testing.ts"
  }
}
