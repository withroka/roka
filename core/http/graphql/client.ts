/**
 * This module provides the {@linkcode client} function to create a
 * GraphQL client for making requests. Useful for building GraphQL-based API
 * clients.
 *
 * ```ts
 * import { client } from "@roka/http/graphql/client";
 * async function usage() {
 *   const api = client("https://api.github.com/graphql");
 *   const query = await Deno.readTextFile("repo.graphql");
 *   const repo = await api.query(query, {
 *     owner: "owner", name: "name",
 *   });
 * }
 * ```
 *
 * @module client
 */

import {
  type AnyVariables,
  cacheExchange,
  Client as UrqlClient,
  fetchExchange,
} from "@urql/core";
import { retryExchange } from "@urql/exchange-retry";
import { RequestError } from "../request.ts";

export { gql } from "@urql/core";

/** A GraphQL client returned by the {@linkcode client} function. */
export interface Client {
  /** Makes a GraphQL query. */
  query<Result>(
    query: string,
    variables?: Record<string, unknown>,
  ): Promise<Result>;
  /** Makes a paginated GraphQL query, given a {@linkcode Paginator}. */
  queryPaginated<Result, Node, Edge, PageInfo>(
    query: string,
    paginator: Paginator<Result, Node, Edge, PageInfo>,
    variables?: Record<string, unknown>,
  ): Promise<Node[]>;
}

/** Options for the {@linkcode client} function. */
export interface ClientOptions {
  /** The bearer token to be sent with the request headers. */
  token?: string;
}

/** A GraphQL paginator that traverses response data to generate pages.
 *
 * A custom paginator is needed by {@linkcode Client.queryPaginated} for
 * nodes it is querying.
 */
export interface Paginator<Result, Node, Edge, PageInfo> {
  /** Extracts edges from the response data. */
  edges(data: Result): Edge[];
  /** Extracts nodes from an edge. */
  node(edge: Edge): Node;
  /** Extracts an object that contains page and cursor information. */
  pageInfo(data: Result): PageInfo;
  /**
   * Extracts the cursor for the next page from page information.
   *
   * This function must return `null` if there are no more pages.
   */
  cursor(pageInfo: PageInfo): string | null;
}

/**
 * Creates an HTTP client for making GraphQL requests.
 *
 * @example Make a GraphQL query.
 * ```ts
 * import { client } from "@roka/http/graphql/client";
 *
 * interface Repository {
 *   owner: string;
 *   name: string;
 *   description: string;
 * }
 *
 * async function usage() {
 *   const api = client("https://api.github.com/graphql");
 *   const { repository } = await api.query<{ repository: Repository }>(`
 *     query($owner: String!, $name: String!) {
 *       repository(owner: $owner, name: $name) {
 *         description
 *       }
 *    }
 *  `,
 *  { owner: "owner", name: "repo" });
 *  console.log(repository.description);
 * };
 * ```
 *
 * @example Make a paginated GraphQL query.
 * ```ts
 * import { client } from "@roka/http/graphql/client";
 *
 * interface Issue {
 *  number: number;
 *  title: string;
 *  state: string;
 * }
 *
 * interface Issues {
 *   repository: {
 *     issues: {
 *       nodes: Issue[];
 *       pageInfo: {
 *         endCursor: string | null;
 *         hasNextPage: boolean;
 *       };
 *     };
 *   };
 * }
 *
 * async function usage() {
 *   const api = client("https://api.github.com/graphql");
 *   const issues: Issue[] = await api.queryPaginated(`
 *     query($owner: String!, $name: String!) {
 *        repository(owner: "owner", name: $name) {
 *          issues(first: 1) {
 *            nodes {
 *              number
 *              state
 *            }
 *          }
 *          pageInfo {
 *            endCursor
 *            startCursor
 *            hasNextPage
 *            hasPreviousPage
 *          }
 *        }
 *      }
 *  `,
 *  {
 *    edges: (issues: Issues) => issues.repository.issues.nodes,
 *    node: (edge) => edge,
 *    pageInfo: (issues) => issues.repository.issues.pageInfo,
 *    cursor: (pageInfo) => pageInfo.hasNextPage ? pageInfo.endCursor : null,
 *  },
 *  { owner: "owner", name: "name" });
 *  console.log(issues);
 * }
 * ```
 */
export function client(url: string | URL, options?: ClientOptions): Client {
  if (typeof url === "string") url = new URL(url);
  const client = new UrqlClient({
    url: url.toString(),
    exchanges: [cacheExchange, retryExchange({}), fetchExchange],
    fetchOptions: () => {
      return {
        headers: {
          ...options?.token && { "Authorization": `Bearer ${options.token}` },
        },
      };
    },
  });
  return {
    async query<Result>(
      query: string,
      variables: Record<string, unknown> = {},
    ): Promise<Result> {
      const response = await client.query(query, variables);
      if (response.error) {
        throw new RequestError(response.error.message, {
          status: response.data?.status,
          cause: response.error,
        });
      }
      return response.data as Result;
    },
    async queryPaginated<Result, Node, Edge, PageInfo>(
      query: string,
      paginator: Paginator<Result, Node, Edge, PageInfo>,
      variables: AnyVariables & { cursor?: string; limit?: number } = {},
    ): Promise<Node[]> {
      const nodes: Node[] = [];
      let cursor: string | null = null;
      do {
        // deno-lint-ignore no-await-in-loop
        const data = await this.query<Result>(query, { ...variables, cursor });
        const edges = paginator.edges(data);
        nodes.push(...edges.map(paginator.node));
        const pageInfo = paginator.pageInfo(data);
        cursor = paginator.cursor(pageInfo);
      } while (
        cursor &&
        (variables.limit === undefined || nodes.length < variables.limit)
      );
      return nodes;
    },
  };
}
