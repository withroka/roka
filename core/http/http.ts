/**
 * A library with helpers for making HTTP requests.
 *
 * This package provides convenience utilities for making HTTP requests and
 * testing requests with mocks.
 *
 * The {@linkcode [request].request | request} function is the core of the
 * package, providing a wrapper around the `fetch` API, while handling errors
 * and retries.
 *
 * ```ts
 * import { request } from "@roka/http/request";
 * async function  usage() {
 *   const response = await request("https://www.example.com", {
 *     method: "GET",
 *     retry: { maxAttempts: 2 },
 *   });
 * }
 * ```
 *
 * The {@link [json]} and {@link [graphql]} modules provide higher-level
 * abstractions for making JSON and GraphQL requests, respectively.
 *
 * ```ts
 * import { client } from "@roka/http/json/client";
 * async function usage() {
 *   const api = client("https://www.example.com");
 *   const data = await api.get<{ id: number; }>("/api/path");
 * }
 * ```
 *
 * ## Modules
 *
 *  -  {@link [request]}: Make simple HTTP requests.
 *  -  {@link [json/client]}: Make JSON requests.
 *  -  {@link [graphql/client]}: Make GraphQL requests.
 *  -  {@link [testing]}: Test with mock requests.
 *
 * @module http
 */
