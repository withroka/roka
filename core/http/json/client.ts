/**
 * This module provides the {@linkcode client} function to make JSON
 * requests and receive JSON responses. It is useful for building JSON-based
 * API clients.
 *
 * ```ts
 * import { client } from "@roka/http/json/client";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = client("https://api.github.com");
 *   await api.post<Issue>(
 *     "/repos/owner/repo/issues",
 *     { body: { title: "Test issue" } },
 *   );
 *   const issue = await api.get<Issue>("/repos/owner/repo/issues/1");
 * }
 * ```
 *
 * @module client
 */

import { request, type RequestOptions } from "../request.ts";

/** A JSON client returned by the {@linkcode client} function. */
export interface Client {
  /** Makes a GET request. */
  get<T>(
    path: string,
    options?: JsonRequestOptions,
  ): Promise<Partial<T>>;
  /** Makes a DELETE request. */
  delete<T>(
    path: string,
    options?: JsonRequestOptions,
  ): Promise<Partial<T>>;
  /** Makes a PATCH request. */
  patch<T>(
    path: string,
    options?: JsonRequestOptions & { body?: object },
  ): Promise<Partial<T>>;
  /** Makes a POST request. */
  post<T>(
    path: string,
    options?: JsonRequestOptions & { body?: object },
  ): Promise<Partial<T>>;
  /** Makes a PUT request. */
  put<T>(
    path: string,
    options?: JsonRequestOptions & { body?: object },
  ): Promise<Partial<T>>;
  /**
   * Makes a GET request with optional JSON body and returns the response as a
   * string.
   */
  text(
    path: string,
    options?: JsonRequestOptions & { body?: object },
  ): Promise<string>;
}

/** Options for JSON requests. */
export type JsonRequestOptions = Omit<RequestOptions, "method" | "body">;

/**
 * Creates an HTTP client for making JSON-based requests.
 *
 * @example Make a JSON request.
 * ```ts
 * import { client } from "@roka/http/json/client";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = client("https://api.github.com");
 *   const issue = await api.get<Issue>("/repos/owner/repo/issues/1");
 * }
 * ```
 *
 * @example Make an authenticated JSON request.
 * ```ts
 * import { client } from "@roka/http/json/client";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = client("https://api.github.com", {
 *     token: "TOKEN",
 *   });
 *   await api.post<Issue>(
 *     "/repos/owner/repo/issues",
 *     { body: { title: "Test issue" } },
 *   );
 * }
 * ```
 */
export function client(
  url: string | URL,
  options?: JsonRequestOptions,
): Client {
  const clientOptions = options;
  if (typeof url === "string") url = new URL(url);
  async function req<T>(
    method: string,
    path: string,
    options: JsonRequestOptions & { body?: object; text?: boolean } = {},
  ): Promise<Partial<T>> {
    const { body, text, ...requestOptions } = options ?? {};
    const response = await request(new URL(path, url), {
      method,
      headers: {
        "Accept": "application/json; charset=UTF-8",
        ...body && { "Content-Type": "application/json; charset=UTF-8" },
      },
      ...clientOptions,
      ...requestOptions,
      ...body && { body: JSON.stringify(body) },
    });
    if (response.headers.get("Content-Type") === null) return {};
    return text ? await response.text() as T : await response.json();
  }
  return {
    get: async (path, options) => await req("GET", path, options),
    delete: async (path, options) => await req("DELETE", path, options),
    patch: async (path, options) => await req("PATCH", path, options),
    put: async (path, options) => await req("PUT", path, options),
    post: async (path, options) => await req("POST", path, options),
    text: async (path, options) =>
      await req<string>("GET", path, { ...options, text: true }),
  };
}
