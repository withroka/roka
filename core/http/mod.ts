/**
 * Helpers for making HTTP requests.
 *
 * The `request` module provides helpers for making HTTP requests. It provides
 * the {@linkcode request} function, which is a wrapper around the Fetch API
 *
 * The `json` module provides the basics for building JSON based API clients.
 * It provides the {@linkcode client} function, which can make authenticated
 * JSON requests.
 *
 * The `graphql` module provides the basics for building GraphQL clients. It
 * provides the {@linkcode client} function, which can make authenticated
 * GraphQL requests.
 *
 * @module
 */

export * from "./request.ts";
