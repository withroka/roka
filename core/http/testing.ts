/**
 * This module provides the {@linkcode mockFetch} function to create a mock for
 * the global `fetch` function. The mocked fetch function can record and replay
 * all global `fetch` calls in tests.
 *
 * ```ts
 * import { mockFetch } from "@roka/http/testing";
 * Deno.test("mockFetch()", async (t) => {
 *   using fetch = mockFetch(t);
 *   await fetch("https://example.com");
 * });
 * ```
 *
 * The mocking system mimics the behavior of the `@std/testing/snapshot`
 * module. Running tests with the `--update` or `-u` flag will create a mock
 * file in the `__mocks__` directory, using real calls. The mock file will be
 * used in subsequent test runs when these flags are not present.
 *
 * When using the mock, the `--allow-read` permission must be enabled, or else
 * any calls will fail due to insufficient permissions. Additionally, when
 * updating the mock, the `--allow-write` and `--allow-net` permissions must be
 * enabled.
 *
 * @module testing
 */

import { type Mock, mock, type MockOptions } from "@roka/testing/mock";
import { pick } from "@std/collections";

export type { Mock, MockOptions } from "@roka/testing/mock";

/** Options for the {@linkcode mockFetch} function. */
export interface MockFetchOptions extends MockOptions {
  /** Limit what gets matched with replay calls. */
  ignore?: {
    /**
     * Whether to ignore the request headers when matching `fetch` calls.
     *
     * Authentication headers (`Authorization`, `Cookie`, `Set-Cookie`) are
     * never stored in mock files and matched, regardless of this option.
     *
     * @default {false}
     */
    headers?: true;
  };
}

/**
 * Create a mock for the global `fetch` function.
 *
 * The behavior of this mock follows the same for any mock created with the
 * `@roka/testing/mock` module. Running tests with the `--update` flag will
 * create a mock file in the `__mocks__` directory, using real fetch calls. The
 * mock file will be used in subsequent test runs when the flag is not present.
 *
 * In `"replay"` mode, a call to the mock will throw {@linkcode MockError} if
 * no matching call was recorded. At the end of the test, the mock will verify
 * that at least one call was made, and all recorded calls were replayed.
 *
 * @example Mock a fetch call.
 * ```ts
 * import { mockFetch } from "@roka/http/testing";
 * import { assertEquals } from "@std/assert";
 *
 * Deno.test("mockFetch()", async (t) => {
 *   using fetch = mockFetch(t);
 *   const response = await fetch("https://example.com");
 *   assertEquals(response.status, 200);
 * });
 * ```
 *
 * @param context The test context.
 * @param options Options for the mock.
 * @return A mock function that records and replays calls to the global `fetch`.
 */
export function mockFetch(
  context: Deno.TestContext,
  options?: MockFetchOptions,
): Mock<typeof fetch> {
  return mock(context, globalThis, "fetch", {
    conversion: {
      input: {
        async convert(input: RequestInfo | URL, init?: RequestInit) {
          const request = input instanceof Request
            ? input.clone()
            : new Request(input, init);
          request.headers.delete("Authorization");
          request.headers.delete("Cookie");
          request.headers.delete("Set-Cookie");
          let body = request.body === null ? undefined : await request.text();
          const contentType = request.headers.get("Content-Type");
          const boundary = contentType?.match(/boundary=(\S+)/)?.[1];
          if (body && boundary) {
            body = body.replaceAll(boundary, "BOUNDARY");
            request.headers.set(
              "Content-Type",
              contentType.replace(boundary, "BOUNDARY"),
            );
          }
          const headers = options?.ignore?.headers
            ? undefined
            : stripHeaders(request.headers);
          return [request.url, {
            ...body && { body },
            ...headers && { headers },
            ...pick(request, [
              "cache",
              "credentials",
              "integrity",
              "keepalive",
              "method",
              "mode",
              "redirect",
              "referrer",
              "referrerPolicy",
            ]),
          }];
        },
      },
      output: {
        async convert(response): Promise<[string | undefined, ResponseInit]> {
          response = response.clone();
          const body = response.body === null
            ? undefined
            : await response.text();
          const headers = stripHeaders(response.headers);
          const init = {
            ...headers && { headers },
            ...pick(response, ["status", "statusText"]),
          };
          return [body, init];
        },
        revert(data) {
          return new Response(...data);
        },
      },
    },
    ...options,
  });
}

function stripHeaders(headers: HeadersInit): HeadersInit {
  const stripped = new Headers(headers);
  stripped.delete("Authorization");
  stripped.delete("Cookie");
  stripped.delete("Set-Cookie");
  return Object.fromEntries(stripped.entries());
}
