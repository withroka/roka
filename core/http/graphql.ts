/**
 * This module provides the {@linkcode graphqlClient} function to create a GraphQL
 * client for making requests. Useful for building GraphQL based API clients.
 *
 * ```ts
 * import { graphqlClient } from "@roka/http/graphql";
 * async function usage() {
 *   const api = graphqlClient("https://api.github.com/graphql");
 *   const query = await Deno.readTextFile("repo.graphql");
 *   const repo = await api.query(query, {
 *     owner: "owner", name: "name",
 *   });
 * }
 * ```
 *
 * @module graphql
 */

import { RequestError } from "@roka/http/request";
import {
  type AnyVariables,
  cacheExchange,
  Client,
  fetchExchange,
} from "@urql/core";
import { retryExchange } from "@urql/exchange-retry";

/** A GraphQL client returned by {@linkcode graphqlClient}. */
export interface GraphQLClient {
  /** Makes a GraphQL query. */
  query<Result>(
    query: string,
    variables?: Record<string, unknown>,
  ): Promise<Result>;
  /** Makes a paginated GraphQL query, given a {@linkcode GraphQLPaginator}. */
  queryPaginated<Result, Node, Edge, PageInfo>(
    query: string,
    paginator: GraphQLPaginator<Result, Node, Edge, PageInfo>,
    variables?: Record<string, unknown>,
  ): Promise<Node[]>;
}

/** Options for {@linkcode graphqlClient}. */
export interface GraphQLClientOptions {
  /** The bearer token to be sent with the request headers. */
  token?: string;
}

/** A GraphQL paginator that traverses response data to generate pages.
 *
 * A custom paginator is needed by {@linkcode graphqlClient.queryPaginated} for
 * nodes it is querying.
 */
export interface GraphQLPaginator<Result, Node, Edge, PageInfo> {
  /** Extracts edges from the response data. */
  edges(data: Result): Edge[];
  /** Extracts nodes from an edge. */
  node(edge: Edge): Node;
  /** Extracts an object that contains page and cursor information. */
  pageInfo(data: Result): PageInfo;
  /**
   * Extracts the cursor for the next page from page information.
   *
   * This function must return `null` if there are no more pages.
   */
  cursor(pageInfo: PageInfo): string | null;
}

/**
 * Creates an HTTP client for making GraphQL requests.
 *
 * @example Make a GraphQL query.
 * ```ts
 * import { graphqlClient } from "@roka/http/graphql";
 *
 * interface Repository {
 *   owner: string;
 *   name: string;
 *   description: string;
 * }
 *
 * async function usage() {
 *   const api = graphqlClient("https://api.github.com/graphql");
 *   const { repository } = await api.query<{ repository: Repository }>(`
 *     query($owner: String!, $name: String!) {
 *       repository(owner: $owner, name: $name) {
 *         description
 *       }
 *    }
 *  `,
 *  { owner: "owner", name: "repo" });
 *  console.log(repository.description);
 * };
 * ```
 *
 * @example Make a paginated GraphQL query.
 * ```ts
 * import { graphqlClient } from "@roka/http/graphql";
 *
 * interface Issue {
 *  number: number;
 *  title: string;
 *  state: string;
 * }
 *
 * interface Issues {
 *   repository: {
 *     issues: {
 *       nodes: Issue[];
 *       pageInfo: {
 *         endCursor: string | null;
 *         hasNextPage: boolean;
 *       };
 *     };
 *   };
 * }
 *
 * async function usage() {
 *   const api = graphqlClient("https://api.github.com/graphql");
 *   const issues: Issue[] = await api.queryPaginated(`
 *     query($owner: String!, $name: String!) {
 *        repository(owner: "owner", name: $name) {
 *          issues(first: 1) {
 *            nodes {
 *              number
 *              state
 *            }
 *          }
 *          pageInfo {
 *            endCursor
 *            startCursor
 *            hasNextPage
 *            hasPreviousPage
 *          }
 *        }
 *      }
 *  `,
 *  {
 *    edges: (issues: Issues) => issues.repository.issues.nodes,
 *    node: (edge) => edge,
 *    pageInfo: (issues) => issues.repository.issues.pageInfo,
 *    cursor: (pageInfo) => pageInfo.hasNextPage ? pageInfo.endCursor : null,
 *  },
 *  { owner: "owner", name: "name" });
 *  console.log(issues);
 * }
 * ```
 */
export function graphqlClient(
  url: string,
  options?: GraphQLClientOptions,
): GraphQLClient {
  const client = new Client({
    url,
    exchanges: [cacheExchange, retryExchange({}), fetchExchange],
    fetchOptions: () => {
      return {
        headers: {
          ...options?.token && { "Authorization": `Bearer ${options.token}` },
        },
      };
    },
  });
  return {
    async query<Result>(
      query: string,
      variables: Record<string, unknown> = {},
    ): Promise<Result> {
      const response = await client.query(query, variables);
      if (response.error) {
        throw new RequestError(response.error.message, response.data.status);
      }
      return response.data as Result;
    },
    async queryPaginated<Result, Node, Edge, PageInfo>(
      query: string,
      paginator: GraphQLPaginator<Result, Node, Edge, PageInfo>,
      variables: AnyVariables & { cursor?: string; limit?: number } = {},
    ): Promise<Node[]> {
      const nodes: Node[] = [];
      let cursor: string | null = null;
      do {
        // deno-lint-ignore no-await-in-loop
        const data = await this.query<Result>(query, { ...variables, cursor });
        const edges = paginator.edges(data);
        nodes.push(...edges.map(paginator.node));
        const pageInfo = paginator.pageInfo(data);
        cursor = paginator.cursor(pageInfo);
      } while (
        cursor &&
        (variables.limit === undefined || nodes.length < variables.limit)
      );
      return nodes;
    },
  };
}
