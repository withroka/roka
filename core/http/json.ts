/**
 * This module provides the {@linkcode jsonClient} function to make JSON requests
 * and receive JSON responses. It is useful for building JSON based API clients.
 *
 * ```ts
 * import { jsonClient } from "@roka/http/json";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = jsonClient("https://api.github.com");
 *   await api.post<Issue>("/repos/owner/repo/issues", {
 *     title: "Test issue",
 *   });
 *   const issue = await api.get<Issue>("/repos/owner/repo/issues/1");
 * }
 * ```
 *
 * @module json
 */

import { request } from "@roka/http/request";

/** A JSON client returned by {@linkcode jsonClient}. */
export interface JsonClient {
  /** Makes a GET request. */
  get<T>(path: string): Promise<Partial<T>>;
  /** Makes a DELETE request. */
  delete<T>(path: string): Promise<Partial<T>>;
  /** Makes a PATCH request. */
  patch<T>(path: string, body: object): Promise<Partial<T>>;
  /** Makes a POST request. */
  post<T>(path: string, body: object): Promise<Partial<T>>;
  /** Makes a PUT request. */
  put<T>(path: string, body: object): Promise<Partial<T>>;
}

/** Options for {@linkcode jsonClient}. */
export interface JsonClientOptions {
  /** The bearer token to be sent with the request headers. */
  token?: string;
  /** The user agent to be sent with the request headers. */
  agent?: string;
  /** The referrer to be sent with the request headers. */
  referrer?: string;
}

/**
 * Creates an HTTP client for making JSON-based requests.
 *
 * @example Make a JSON request.
 * ```ts
 * import { jsonClient } from "@roka/http/json";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = jsonClient("https://api.github.com");
 *   const issue = await api.get<Issue>("/repos/owner/repo/issues/1");
 * }
 * ```
 *
 * @example Make an authenticated JSON request.
 * ```ts
 * import { jsonClient } from "@roka/http/json";
 * async function usage() {
 *   interface Issue { number: number; state: string; };
 *   const api = jsonClient("https://api.github.com", {
 *     token: "TOKEN",
 *   });
 *   await api.post<Issue>("/repos/owner/repo/issues", {
 *     title: "Test issue",
 *   });
 * }
 * ```
 */
export function jsonClient(
  url: string,
  options?: JsonClientOptions,
): JsonClient {
  async function req<T>(
    method: string,
    path: string,
    body?: object,
  ): Promise<Partial<T>> {
    const response = await request(`${url}${path}`, {
      method,
      headers: {
        "Accept": "application/json; charset=UTF-8",
        ...body && { "Content-Type": "application/json; charset=UTF-8" },
        ...(options?.token && { "Authorization": `Bearer ${options.token}` }),
        ...(options?.agent && { "User-Agent": options.agent }),
        ...(options?.referrer && { "Referrer": options.referrer }),
      },
      ...body && { body: JSON.stringify(body) },
    });
    if (response.headers.get("Content-Type") === null) return {};
    return await response.json();
  }
  return {
    async get(path) {
      return await req("GET", path);
    },
    async delete(path) {
      return await req("DELETE", path);
    },
    async patch(path, body) {
      return await req("PATCH", path, body);
    },
    async put(path, body) {
      return await req("PUT", path, body);
    },
    async post(path, body) {
      return await req("POST", path, body);
    },
  };
}
