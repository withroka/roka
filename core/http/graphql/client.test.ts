import { mockFetch } from "@roka/http/testing";
import { assertSnapshot } from "@std/testing/snapshot";
import { client, gql } from "./client.ts";

const token = Deno.env.get("GITHUB_TOKEN") ?? "TOKEN";

interface Repository {
  owner: string;
  name: string;
  description: string;
}

interface Issue {
  number: number;
  title: string;
  state: string;
}

interface Issues {
  repository: {
    issues: {
      nodes: Issue[];
      pageInfo: {
        endCursor: string | null;
        hasNextPage: boolean;
      };
    };
  };
}

Deno.test("client().query() makes GraphQL query", async (t) => {
  using _fetch = mockFetch(t);
  const api = client("https://api.github.com/graphql", { token });
  const result: Repository = await api.query(
    gql`
      query($owner: String!, $name: String!) {
        repository(owner: $owner, name: $name) {
          owner {
            login
          },
          name,
          description
        }
      }
    `,
    { owner: "withroka", name: "test" },
  );
  await assertSnapshot(t, result);
});

Deno.test("client().queryPaginated() makes paginated GraphQL query", async (t) => {
  using _fetch = mockFetch(t);
  const api = client("https://api.github.com/graphql", { token });
  const result = await api.queryPaginated(
    gql`
      query($owner: String!, $name: String!) {
        repository(owner: $owner, name: $name) {
          issues(first: 100) {
            nodes {
              number
              title
              state
            }
            pageInfo {
              endCursor
              startCursor
              hasNextPage
              hasPreviousPage
            }
          }
        }
      }
    `,
    {
      edges: (issues: Issues) => issues.repository.issues.nodes,
      node: (edge) => edge,
      pageInfo: (issues) => issues.repository.issues.pageInfo,
      cursor: (pageInfo) => pageInfo.hasNextPage ? pageInfo.endCursor : null,
    },
    { owner: "withroka", name: "test" },
  );
  await assertSnapshot(t, result);
});
