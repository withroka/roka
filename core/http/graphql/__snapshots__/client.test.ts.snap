export const snapshot = {};

snapshot[`client().query() makes GraphQL query 1`] = `
{
  repository: {
    __typename: "Repository",
    description: "Dummy repository for E2E testing",
    name: "test",
    owner: {
      __typename: "Organization",
      login: "withroka",
    },
  },
}
`;

snapshot[`client().queryPaginated() makes paginated GraphQL query 1`] = `
[
  {
    __typename: "Issue",
    number: 7,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 8,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 9,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 10,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 11,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 12,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 13,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 14,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 15,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 16,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 17,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 18,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 19,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 20,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 21,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 22,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 23,
    state: "CLOSED",
    title: "Test",
  },
  {
    __typename: "Issue",
    number: 24,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 45,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 47,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 48,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 49,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 52,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 54,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 55,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 56,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 57,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 58,
    state: "CLOSED",
    title: "Test issue",
  },
  {
    __typename: "Issue",
    number: 59,
    state: "CLOSED",
    title: "Test issue",
  },
]
`;
